src/relayserver/ReputationStoreManager.ts

```diff
@@ -0,0 +1,89 @@
+import AsyncNedb from 'nedb-async'
+
+import { LoggerInterface } from '../common/LoggerInterface'
+import { ReputationChange, ReputationEntry } from './ReputationEntry'
+import { Address } from '../relayclient/types/Aliases'
+
+export const REPUTATION_STORE_FILENAME = 'reputation_store.db'
+
+export class ReputationStoreManager {
+  private readonly txstore: AsyncNedb<any>
+  private readonly logger: LoggerInterface
+
+  constructor ({ workdir = '/tmp/test/' }, logger: LoggerInterface) {
+    this.logger = logger
+    const filename = `${workdir}/${REPUTATION_STORE_FILENAME}`
+    this.txstore = new AsyncNedb({
+      filename,
+      autoload: true,
+      timestampData: true
+    })
+    this.txstore.ensureIndex({ fieldName: 'paymaster', unique: true })
+    this.logger.info(`Reputation system database location: ${filename}`)
+  }
+
+  async createEntry (paymaster: Address, reputation: number): Promise<ReputationEntry> {
+    const entry: ReputationEntry = {
+      paymaster: paymaster.toLowerCase(),
+      reputation,
+      lastAcceptedRelayRequestTs: 0,
+      abuseStartedTs: 0,
+      changes: []
+    }
+    return await this.txstore.asyncInsert(entry)
+  }
+
+  async clearAbuseFlag (paymaster: Address, reputation: number): Promise<void> {
+    const update: Partial<ReputationEntry> = {
+      reputation,
+      abuseStartedTs: 0
+    }
+    await this.updateEntry(paymaster, update)
+  }
+
+  async setAbuseFlag (paymaster: Address): Promise<void> {
+    const update: Partial<ReputationEntry> = {
+      abuseStartedTs: Date.now()
+    }
+    this.logger.warn(`Paymaster ${paymaster} was flagged as abused`)
+    await this.updateEntry(paymaster, update)
+  }
+
+  async updateLastAcceptedTimestamp (paymaster: Address): Promise<void> {
+    const lastAcceptedRelayRequestTs = Date.now()
+    const update: Partial<ReputationEntry> = {
+      lastAcceptedRelayRequestTs
+    }
+    this.logger.debug(`Paymaster ${paymaster} was last accepted at ${lastAcceptedRelayRequestTs}`)
+    await this.updateEntry(paymaster, update)
+  }
+
+  async updatePaymasterReputation (paymaster: Address, change: number): Promise<void> {
+    const existing: ReputationEntry = await this.txstore.asyncFindOne({ paymaster: paymaster.toLowerCase() })
+    const reputationChange: ReputationChange = {
+      timestamp: Date.now(),
+      change
+    }
+    const reputation = existing.reputation + change
+    const update: Partial<ReputationEntry> = {
+      reputation,
+      changes: [...existing.changes, reputationChange]
```

Comment:
its a persistent store, and you keep appending it.. better filter out old entries (e.g. week old)

Reply:
Fixed. Not explicitly tested as I kept it at a database layer and I want to switch the entire database layer as we are using an long-abandoned inconsistently-typed file-writing library.